### PR TITLE
Llm streamlit

### PR DESCRIPTION
**Pull request by YajieLi-Z**
Implement the LLM Version of the Streamlit App with GPT-4 Integration

Features:
- Generate a JSON format from the table output by GPT-4 that can be pushed to DHIS2.
- Add a configure_openai function similar to the DHIS2 configuration.
- Update prompts for improved performance.

Checklist:
Please check the README file and ensure the OpenAI API key is included.

How to Run:
Install the required packages:
pip install -r requirements.txt

Run the Streamlit app:
streamlit run app_llm.py

### CODE REVIEW

File: LLM/ocr_functions.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        exif = dict(image.getexif().items())
+        if exif.get(orientation) == 3:
+            image = image.rotate(180, expand=True)
+        elif exif.get(orientation) == 6:
+            image = image.rotate(270, expand=True)
+        elif exif.get(orientation) == 8:
+            image = image.rotate(90, expand=True)
+    except (AttributeError, KeyError, IndexError):
+        pass
+    return image
```
Comment (ginic) [inline LLM/ocr_functions.py:153]: When working with any files in python, you always want to use a `with` statement or a try...close paradigm, which will close the file and prevent memory leaks. See https://realpython.com/python-with-statement/ for more details. For example: 
```suggestion
    with Image.open(image_path) as image:
      orientation = None
      try:
          for orientation in ExifTags.TAGS.keys():
              if ExifTags.TAGS[orientation] == 'Orientation':
                  break
          exif = dict(image.getexif().items())
          if exif.get(orientation) == 3:
              image = image.rotate(180, expand=True)
          elif exif.get(orientation) == 6:
              image = image.rotate(270, expand=True)
          elif exif.get(orientation) == 8:
              image = image.rotate(90, expand=True)
      except (AttributeError, KeyError, IndexError):
          pass
      return image.copy()
```

File: LLM/ocr_functions.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
+import base64
+import json
+import pandas as pd
+from PIL import Image, ExifTags
+from openai import OpenAI
+from msfocr.data import data_upload_DHIS2 as dhis2
+import configparser
+
+
```
Comment (ginic) [inline LLM/ocr_functions.py:10]: This isn't necessary for the code to work, but I like to declare my module constants all together at the top of the module, even if they aren't set in the code until later. This makes it easier to read the code and understand what's going on and what variables are important. 

```suggestion
OPENAI_API_KEY = None

```

File: README.md [hunk 3]
```diff
@@ -20,5 +20,8 @@ In the main directory:
 username = '<your username here>'
 password = '<your password here>'
 server_url = '<test server url>'
+
+[OpenAI]
+api_key = "your api key here"
```
Comment (ginic) [inline README.md]: Just for consistency, and actually the quotes aren't needed:
```suggestion
api_key=<your api key here>
```

**Review by ginic — CHANGES_REQUESTED** [review]
Thanks, @ByteYJ ! The GPT OCR model works really well for me and I was able to get everything running. 

I noticed a few minor things to change which I included as suggestions in the comments. Please take a look at those. Once those items are addressed, I think it will be ready to merge in!

File: LLM/ocr_functions.py (hunk 1 above)
Comment (YajieLi-Z) [reply LLM/ocr_functions.py:153]: It works in some cases. Using it directly can help eliminate errors?

File: app_llm.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+                    st.write("Period Type\: " + period_type)
+
+    # Initialize with period values recognized from tally sheet or entered by user
+    if (start_date and end_date):
+        if start_date:
+            period_start = st.date_input("Period Start Date", format="YYYY-MM-DD", value=start_date)
+        else:
+            period_start = st.date_input("Period Start Date", format="YYYY-MM-DD")
+        if end_date:
+            period_end = st.date_input("Period End Date", format="YYYY-MM-DD", value=start_date)
```
Comment (anjus1313) [inline app_llm.py]: We can remove all the period_end statements as we only need the periodType and period_start to determine the time period.

**Comment by anjus1313** [comment]
Note for later:

Images of size greater than 4MB are unable to be processed by GPT (even though the error mentions 20MB file size limit). Could be because of encode_image function that processes the image prior to sending it to GPT. Might be worth looking at image compression or resizing techniques.

File: app_llm.py (hunk 4 above)
Comment (YajieLi-Z) [reply app_llm.py]: Thanks! Removed~

**Comment by YajieLi-Z** [comment]
@anjus1313 Thanks for the note. The OpenAI API does its own internal multi-step resizing. It seems that the maximum size of an image is 2048x768, which gets divided into 8 512x512 tiles. Anything larger gets resized to this. The 5.2MB works. T
Screenshot 2024-07-18 at 1 22 36 PM
size. Will figure it out for upload issue reason of streamlit.

**Review by anjus1313 — APPROVED** [review]

**Review by ginic — APPROVED** [review]

**Pull request merged**